### PR TITLE
feat(nim): added nim parser and queries

### PR DESCRIPTION
**Pull request by aMOPel**
### Features
  
  - [x] Highlighting
  - [x] Language Injection
    - `SQL` and `RegEx` strings
    - any language you want in user defined generalized string literals
    - `std/strformat` strings, including debugging `=` and other
      [quirks](https://nim-lang.org/docs/strformat.html#standard-format-specifiers-for-strings-integers-and-floats) (see Notes)
    - `MarkDown` in documentation comments
    - `C/Cpp/ObjC/JavaScript` in emit pragma and `asm` statements (see Notes)
  - [x] Folds
  - [x] Locals

 ### NOTES

#### Format Strings

**Requires** `:TSInstall nim_format_string`.

In format strings `{`/`}` have to be escaped, which breaks the `Nim parser`,
thus expressions in format strings containing `\{`/`\}` will intentionally not be parsed.

#### Emit Pragma / asm statements

Parsing in the emit pragma requires the language name in comment on the preceding line.
E.g.:

```nim
#cpp           <- necessary
{.emit: """
...
<cpp code>
...
""".}

#javascript           <- necessary
asm """
<js code>
"""
```

### CODE REVIEW

File: lua/nvim-treesitter/parsers.lua
Comment (clason) [inline lua/nvim-treesitter/parsers.lua:1134]:
```suggestion
    files = { "src/parser.c", "src/scanner.c" },
```

File: queries/nim/injections.scm
Comment (lucario387) [inline queries/nim/injections.scm:88]: push this offset to the end so that it's easier to read the predicate and query. Also, I think the parser should be doing the heavy lifting and create a string contents node here, and then it's captured for injection

```suggestion
((comment) @injection.language
 . 
 (pragma_statement
   (pragma_list
     (colon_expression
       left: (identifier) @_emit_keyword
       right: (string_literal "\"" . "\"") @injection.content)))
 (#eq? @_emit_keyword "emit")
 (#offset! @injection.language 0 2 0 0)
 (#offset! @injection.content 0 1 0 -1))

```

Comment (aMOPel) [reply queries/nim/injections.scm:88]: @alaviss

File: queries/nim/folds.scm
Comment (lucario387) [inline queries/nim/folds.scm:31]: In this case, is it better to just capture `argument_list` ? Or is the call supposed to be chained as well

Comment (aMOPel) [reply queries/nim/folds.scm:31]: That is about a special case in the syntax, where you can pass whole statements (not expressions) as arguments to a function. It is primarily used for macros and templates.

```nim
template_call(arg1:int, arg2:int):
  # here starts the (statement_list) as part of the (argument_list)
  arg3_statement_1 # here would be real statements
  arg3_statement_2
```

So yeah this is specifically about folding the whole call if there is a (statement_list) in the (argument_list)

File: queries/nim/injections.scm
Comment (alaviss) [reply queries/nim/injections.scm:88]: I can make this happen for strings, but only strings if that works for you.

Splitting comment would make this a non-terminal, which is not very nice in terms of `extras`.

Comment (lucario387) [reply queries/nim/injections.scm:88]: Only strings are more than enough.

I'm also okay with the current version, just that the code is a bit unreadable for me

**Comment by aMOPel** [comment]
I'm gonna try to write indent queries after all now, since alaviss pushed some changes to the parser [just now](https://github.com/alaviss/tree-sitter-nim/pull/53)

**Review by lucario387 — COMMENTED** [review]
These two overlaps with one another. You should add a `(#not-eq? @injection.language "fmt")` to the one above

File: queries/nim/injections.scm
Comment (lucario387) [inline queries/nim/injections.scm:22]: The second query is overlapped with the other two

```suggestion
(generalized_string 
  function: (identifier) @_string_prefix .
  (string_content) @injection.content
  (#set! injection.language "regex")
  (#any-of? @_string_prefix "re" "rex"))

; format string in generalized_strings
(generalized_string 
  function: (identifier) @_string_prefix .
  (string_content) @injection.content
  (#set! injection.language "nim_format_string")
  (#eq? @_string_prefix "fmt"))

; sql in generalized_strings
; and anything you like as long as the function name is the same as the injected language's parser
(generalized_string 
  function: (identifier) @injection.language
  (string_content) @injection.content
  (#not-any-of? @injection.language "re" "rex" "fmt"))
```

Comment (lucario387) [reply queries/nim/injections.scm:22]: Second thought, does injections for generalized strings with languages that's not sql, regex and fmt occur commonly? If not, I suggest to just narrow it to sql only, just like you've done for fmt and regex

Comment (aMOPel) [reply queries/nim/injections.scm:22]: Afaik there are no other injected languages in generalized string in the stdlib. (Beside peg, which doesn't have a tree sitter parser)

But you can always define your own generalized strings and I thought it would be cool to support injection of every language just like that. You don't think it's worth it?

@alaviss what's your opinion on this?

Comment (lucario387) [reply queries/nim/injections.scm:22]: > You don't think it's worth it?

I don't know to be honest. It may be nice, but if it's too niche in usage, I don't think it's worth to support it. Whatever the tree-sitter-nim authors thinks will be correct, I will be going with that opinion as well.

At least, remove the overlaps for the injections. It's never nice for 2 injection queries to overlap with one another.

Comment (aMOPel) [reply queries/nim/injections.scm:22]: yes will do. good catch

Comment (aMOPel) [reply queries/nim/injections.scm:22]: @lucario387 

A question aside:

I have been conservative with what I put into injection queries because of
https://github.com/neovim/neovim/issues/22426
> Currently we run the full injection query on every edit to a buffer.

Technically it would be correct to support a regex like that: `f[mM][tT]` and similar for other ones. But I thought it's not worth the performance cost.
Also look at this [issue](https://github.com/aMOPel/nvim-treesitter-nim/issues/1) which shows some alternative ways to write those generalized string, for which I currently didn't include queries.

Do you think this is a good decision or should I go for more correctness and performance will be fine?

Comment (lucario387) [reply queries/nim/injections.scm]: You can add whatever you think will be better, as long as the experience between keystrokes is not 100ms long

Comment (aMOPel) [reply queries/nim/injections.scm]: how can I measure that?

Comment (aMOPel) [reply queries/nim/injections.scm]: (also that is just subjective testing with my hardware)

File: queries/nim_format_string/highlights.scm
Comment (ObserverOfTime) [inline queries/nim_format_string/highlights.scm:19]: Why `@error`?

File: queries/nim_format_string/injections.scm
Comment (ObserverOfTime) [inline queries/nim_format_string/injections.scm:4]: Formatting.
```suggestion
((matching_curlies (nim_expression !escaped_curly) @injection.content)
  (#set! injection.language "nim"))
```

File: queries/nim/injections.scm
Comment (ObserverOfTime) [inline queries/nim/injections.scm:68]: This should work.
```suggestion
        right: (_ (string_content) @injection.content)))))
```

File: queries/nim/injections.scm
Comment (ObserverOfTime) [inline queries/nim/injections.scm:25]:
```suggestion
  (_ (string_content) @injection.content) 
```

File: queries/nim/injections.scm
Comment (ObserverOfTime) [inline queries/nim/injections.scm:41]:
```suggestion
;   (#set! injection.language "asm"))
```
Soon™

File: queries/nim/highlights.scm
Comment (ObserverOfTime) [inline queries/nim/highlights.scm:8]: Is the priority necessary?

File: queries/nim/highlights.scm
Comment (ObserverOfTime) [inline queries/nim/highlights.scm:45]: Move the `(operator)` capture below instead?

File: queries/nim/injections.scm
Comment (lucario387) [reply queries/nim/injections.scm]: Yes, smoothness is also subjective. Mostly, just benchmark with plenary, or eye test :D

I'm okay with performance issues for now, we can always fix it going forward

File: queries/nim/highlights.scm
Comment (ObserverOfTime) [inline queries/nim/highlights.scm:363]: Maybe `@preproc` for these two?

File: queries/nim/highlights.scm
Comment (ObserverOfTime) [inline queries/nim/highlights.scm:467]:
```suggestion
  (#set! "priority" 98)) ; for parameters in proc_type
```

File: queries/nim/highlights.scm
Comment (ObserverOfTime) [inline queries/nim/highlights.scm:767]:
```suggestion
; NOTE: we can't know what the left identifier is, so better leave it alone
```

File: queries/nim/highlights.scm
Comment (ObserverOfTime) [inline queries/nim/highlights.scm:797]: See https://github.com/nvim-treesitter/nvim-treesitter/pull/5421#issuecomment-1768624719

File: queries/nim/highlights.scm
Comment (ObserverOfTime) [inline queries/nim/highlights.scm:216]: What about [these builtins](https://nim-lang.org/docs/manual.html#procedures-nonoverloadable-builtins)?

File: queries/nim/highlights.scm
Comment (ObserverOfTime) [inline queries/nim/highlights.scm:1]: Please remove the unused blocks.

**Review by ObserverOfTime — CHANGES_REQUESTED** [review]

File: queries/nim/highlights.scm
Comment (lucario387) [reply queries/nim/highlights.scm:767]: Wait why do need priorities for these? Can't they be moved downwards

File: queries/nim_format_string/highlights.scm
Comment (clason) [reply queries/nim_format_string/highlights.scm:19]: Do not use `@error`; that's a purely diagnostic capture that should not be used for actual highlights. Use `@text.danger` for error-type text instead.

Comment (aMOPel) [reply queries/nim_format_string/highlights.scm:19]: alright. gonna go back to `@none` then

Comment (clason) [reply queries/nim_format_string/highlights.scm:19]: What does that do? Why not simply skip the capture?

File: queries/nim/injections.scm
Comment (aMOPel) [reply queries/nim/injections.scm:68]: is that better for performance?

There are other places where I opted for explicit options over using `_` if they are limited.

File: queries/nim_format_string/highlights.scm
Comment (clason) [reply queries/nim_format_string/highlights.scm:19]: (but, yes, `@none` is usually used for string interpolation, if that's what's happening here.)

Comment (aMOPel) [reply queries/nim_format_string/highlights.scm:19]: https://github.com/aMOPel/tree-sitter-nim-format-string#limitations

It is supposed to highlight the nim expression as white or whatever (instead of string color), when we are not injecting the nim parser.

File: queries/nim/highlights.scm
Comment (aMOPel) [reply queries/nim/highlights.scm:796]: huh.

File: queries/nim/highlights.scm
Comment (aMOPel) [reply queries/nim/highlights.scm:216]: nice. I never found that reference. I was arguing with me which builtins (from system.nim) to include and didn't know where to draw the line so I left it completely.

File: queries/nim/injections.scm
Comment (clason) [reply queries/nim/injections.scm:68]: It's less likely to break if nodes are added or removed from the parser, for one thing.

Comment (aMOPel) [reply queries/nim/injections.scm:68]: doesn't feel like a strong argument for me. everything is likely to break, that's why the parser version is pinned.

more generic can be as bad more specific in terms of breakage.

Eg there could be a new node introduced that contains `(string_content)`

Comment (ObserverOfTime) [reply queries/nim/injections.scm:68]: > is that better for performance?

I don't know about performance but it's better for readability.

> Eg there could be a new node introduced that contains `(string_content)`

That would still be injectable, no?

Comment (clason) [reply queries/nim/injections.scm:68]: > doesn't feel like a strong argument for me. everything is likely to break, that's why the parser version is pinned.

Yes, but we still _try_ to update, and the less manual intervention is needed, the better for us and for you (as the language maintainer).

File: queries/nim_format_string/highlights.scm
Comment (clason) [reply queries/nim_format_string/highlights.scm:19]: Then use `@none`.

File: queries/nim/injections.scm
Comment (aMOPel) [reply queries/nim/injections.scm:68]: > That would still be injectable, no?

Yes, but then it could be injected in the wrong places. And then it would be a silent error. This way we get immediate feedback if something is not working as intended

File: queries/nim/highlights.scm
Comment (aMOPel) [reply queries/nim/highlights.scm]: I removed the prio here

File: queries/nim/highlights.scm
Comment (aMOPel) [reply queries/nim/highlights.scm:1]: will do. But would you have thought about the @function.builtin without them? :b

File: queries/nim/injections.scm
Comment (lucario387) [reply queries/nim/injections.scm:68]: In this case, for the query to be injected in the wrong place, it means there has to be a new node with `(string_content)` that will be placed in the same scenario as this. And most likely that node will also be a string-esque that will need to be injected.

File: queries/nim/highlights.scm
Comment (aMOPel) [reply queries/nim/highlights.scm]: Honestly I don't like this list. Put aside that it seems to be out of date, since `definedInScope` doesn't even exist apparently.

Like I said, there are just [sooo many builtin functions](https://nim-lang.org/docs/system.html#12) that could deserve to be captured here.
Eg, why should `sizeof` be in it, but not `typeof`?

Any line drawn would be arbitrary and the toll on the performance would grow and grow with the list of identifiers to check for.

File: queries/nim/injections.scm
Comment (aMOPel) [reply queries/nim/injections.scm:68]: Still. All we gain is subjective readability. But I can do it if you all insist.

**Comment by aMOPel** [comment]
Just so we're on the same page, I'm gonna squash all those tiny extra commits be we merge right?

**Comment by clason** [comment]
You can squash yourself to signal you're done; otherwise we squash-merge.

**Comment by aMOPel** [comment]
Alright. Squash-merge sounds fine. But I'm still gonna do the indents.scm first.

**Comment by aMOPel** [comment]
Is there some places I could put all my test code examples?

Also @clason , is the support for nvim query unit tests done yet?

**Comment by clason** [comment]
> Is there some places I could put all my test code examples?

yes, under `/tests`. but only those that are specific to the queries you changed here; we neither need nor want to duplicate upstream parser tests.

> Also @clason , is the support for nvim query unit tests done yet?

What do you mean?

**Comment by aMOPel** [comment]
> What do you mean?
 I believe a while ago you wrote on the matrix server that there is upstream (tree-sitter) support coming for unit test customization which would enable us to write actually automatic unit tests for nvim queries.

**Comment by aMOPel** [comment]
> yes, under /tests. but only those that are specific to the queries you changed here; we neither need nor want to duplicate upstream parser tests.

I would take the relevant test code files from [here](https://github.com/aMOPel/nvim-treesitter-nim/tree/main/tests), that I use for manual testing of the queries.

**Comment by lucario387** [comment]
> I believe a while ago you wrote on the matrix server that there is upstream (tree-sitter) support coming for unit test customization which would enable us to write actually automatic unit tests for nvim queries.

That requires a new tree-sitter release (or personal compilation of latest master) which will be arriving soon

**Comment by clason** [comment]
> I believe a while ago you wrote on the matrix server that there is upstream (tree-sitter) support coming for unit test customization which would enable us to write actually automatic unit tests for nvim queries.

That would just allow us to replace our home-grown test crate with upstream tree-sitter, making it easier to run locally. It shouldn't make a difference for CI here.

**Comment by clason** [comment]
> > yes, under /tests. but only those that are specific to the queries you changed here; we neither need nor want to duplicate upstream parser tests.
> 
> I would take the relevant test code files from [here](https://github.com/aMOPel/nvim-treesitter-nim/tree/main/tests), that I use for manual testing of the queries.

No, there's no need for that. If you already run these tests in _your_ CI, no need for us to run them twice. We _only_ want tests that are different from upstream (because you needed to adapt the queries).

File: queries/nim/highlights.scm
Comment (aMOPel) [reply queries/nim/highlights.scm]: same goes for `@type.builtin` btw. With that one it would be also much more impactful on performance, because I would need to duplicate all the `@type` queries to check for all the possible identifiers.

**Comment by aMOPel** [comment]
> No, there's no need for that. If you already run these tests in your CI, no need for us to run them a second time. We only want tests that are different from upstream (because you needed to adapt the queries).

I think you misunderstood. That link I posted is from the repo I am gonna archive once I could merge all my queries. That is no upstream repo.

**Comment by clason** [comment]
But the upstream repo has queries, that are presumably tested. Anything that _can_ be tested there, _should_ be tested there. We just cannot run full CI (on every single PR!) for all 230+ parsers that we distribute.

**Comment by aMOPel** [comment]
What upstream repo are you talking about?

File: queries/nim/highlights.scm
Comment (aMOPel) [reply queries/nim/highlights.scm:8]: this priority is necessary, because the capture is most specific 
and will overrule things like (type_expression) which contain (identifier)s

I'm gonna put this in a comment

File: queries/nim/highlights.scm
Comment (aMOPel) [reply queries/nim/highlights.scm:37]: sure

File: queries/nim/highlights.scm
Comment (lucario387) [reply queries/nim/highlights.scm:8]: Then we can move `(identifier)` up, and then add a `(type_expression (identifier) @type)` instead of needing priorities

File: queries/nim/highlights.scm
Comment (aMOPel) [reply queries/nim/highlights.scm:332]: doesn't even make a difference for my theme.

`macro` and `template` share the syntax with the other function keywords though and I am already capturing the identifier following `macro` and `template` keywords as `@function.macro` and the keywords don't appear standalone.

**Comment by clason** [comment]
> What upstream repo are you talking about?

https://github.com/alaviss/tree-sitter-nim, of course.

File: queries/nim/highlights.scm
Comment (ObserverOfTime) [reply queries/nim/highlights.scm:332]: The colour is irrelevant. If these define preprocessor functions, then `@preproc` is appropriate.

File: queries/nim/highlights.scm
Comment (ObserverOfTime) [reply queries/nim/highlights.scm:1]: I actually thought about it from the static/addr highlights.

**Comment by aMOPel** [comment]
> But the upstream repo has queries, that are presumably tested

but there are only rudimentary highlight queries (that have little in common with mine) and basically no tests in there... And those tests mean nothing in the context of nvim queries. My test files are for the nvim captures. Frankly I haven't written the comments in there yet, just the code examples since I couldn't use the unit testing locally anyways.

**Comment by clason** [comment]
If upstream doesn't want your queries (which would be a bit baffling), then here is fine.

File: queries/nim/highlights.scm
Comment (aMOPel) [reply queries/nim/highlights.scm:8]: `@variable` capture is already at the top.

I had it wrong in my mind. the situation is this:

1. since `(type_expression)`s can infinitely nest, we have to capture the `(identifier)`s inside as `@type` with `has-ancestor` directive,
2. to have `@variable` captures not override `@type` captures those need to come before
3. NOW to have `@type` capture not override `@parameter` captures inside `(proc_type)`s, `@type` captures need to be before `@parameter`  captures.

But you're right, I can remove both priorities and push the `@parameter` block behind the `@type` block

**Comment by aMOPel** [comment]
I made a PR to @alaviss a while ago and we settled on putting the queries here only, since they are nvim specific and the tree sitter cli tests simply don't work correctly, on my queries since I use stuff like `has-ancestor` (and used to use `set priority`)

File: queries/nim/highlights.scm
Comment (lucario387) [inline queries/nim/highlights.scm:389]: This will be costly, much more than most queries in this file

Comment (aMOPel) [reply queries/nim/highlights.scm:389]: you're talking about the has-ancestor?

Comment (lucario387) [reply queries/nim/highlights.scm]: Yep
You basically are traversing the entire depth of the tree with (not-has-ancestor), so this may lead to slowdowns

File: queries/nim/highlights.scm
Comment (lucario387) [inline queries/nim/highlights.scm:425]: I know this is to workaround tree-sitter/tree-sitter#880 but I don't think it's acceptable for nvim-treesitter

Comment (aMOPel) [reply queries/nim/highlights.scm:425]: Why not? What else am I supposed to do?

Comment (clason) [reply queries/nim/highlights.scm:425]: Nothing for now, and wait until tree-sitter has figured this out.

File: queries/nim/highlights.scm
Comment (aMOPel) [reply queries/nim/highlights.scm]: I knew that. I was just being a little dumb, I benchmarked this and it looked fine, but my benchmarking was done with the ts cli, which doesn't know the has-ancestor directive.

File: queries/nim/highlights.scm
Comment (aMOPel) [reply queries/nim/highlights.scm]: @lucario387 have a look at the latest commit, if you like that better

Comment (lucario387) [reply queries/nim/highlights.scm]: Guess I have to pick a poison, priority or deep nesting 😅 

I prefer the former than the latter here.

@clason will decide.

Comment (aMOPel) [reply queries/nim/highlights.scm]: Correctness wise, this solution is even a little better since it catches everything no matter how deep.
Also it got rid of the the performance hungry `has-ancestor` and `not-has-ancestor` call on every `(identifier)`

Comment (clason) [reply queries/nim/highlights.scm]: Priority is tricky to get right because of injections. You may have to tweak the exact values. 

But avoiding hacky nestings or expensive `ancestor` calls are more important.

Comment (aMOPel) [reply queries/nim/highlights.scm]: Tweak how? My injection queries look fine.

Comment (clason) [reply queries/nim/highlights.scm]: You never know if someone doesn't inject Nim into a codeblock somewhere.

Comment (lucario387) [reply queries/nim/highlights.scm]: At least make sure of this one injection: Nim codeblocks in Markdown files have correct highlighting

**Comment by aMOPel** [comment]
I think something is wrong with the CI, since some cpp unit tests seems to be failing

File: queries/nim/highlights.scm
Comment (aMOPel) [reply queries/nim/highlights.scm]: was indeed not working but I think I fixed it

File: queries/nim/indents.scm
Comment (lucario387) [inline queries/nim/indents.scm:1]: Can I get some tests for indents as well

Comment (aMOPel) [reply queries/nim/indents.scm:1]: I removed indents again. `indents.scm` just annoys me like crazy. There are too many cases and I have to manually test every single one. Not willing to do that. Also the queries are just super weird sometimes and I can't figure out why the indent or dedent won't happen. On top of that [alaviss's plugin](https://github.com/alaviss/nim.nvim) already does a good enough job at indenting.

**Comment by aMOPel** [comment]
Alright. If you got nothing else. I'm ready to merge

**Review by ObserverOfTime — APPROVED** [review]

**Review by lucario387 — APPROVED** [review]

**Comment by lucario387** [comment]
Here we go. 

Ty @alaviss and @aMOPel for the parser, queries and changes as per requests

**Pull request merged**
